lib/hound/metadata.ex

```diff
@@ -0,0 +1,96 @@
+defmodule Hound.Metadata do
+  @moduledoc """
+  Metadata allows to pass and extract custom data through.
+  This can be useful if you need to identify sessions.
+
+  The keys and values must be serializable using `:erlang.term_to_binary/1`.
+
+  ## Examples
+
+  You can start a session using metadata by doing the following:
+
+      Hound.start_session(metadata: %{pid: self()})
+
+
+  If you need to retrieve the metadata, you simply need to use
+  `Hound.Metadata.extract/1` on the user agent string, so supposing you are using plug,
+
+
+      user_agent =  conn |> get_resp_header("user-agent") |> List.first
+      metadata   = Hound.Metadata.extract(user_agent)
+      assert %{pid: pid} = metadata
+      # you can use your pid here
+
+  """
+
+  @extract_regexp ~r{HoundMetadata \((.*?)\)}
+
+  @doc """
+  Appends the metdata to the user_agent string.
+  """
+  @spec append(String.t, nil | map | String.t) :: String.t
+  def append(user_agent, nil), do: user_agent
+  def append(user_agent, metadata) when is_map(metadata) or is_list(metadata) do
+    append(user_agent, format(metadata))
+  end
+  def append(user_agent, metadata) when is_binary(metadata) do
+    "#{user_agent}/#{metadata}"
+  end
+
+  @doc """
+  Formats a string to a valid UserAgent string to be passed to be
+  appended to the browser user agent.
+  """
+  @spec format(map | Keyword.t) :: String.t
+  def format(metadata) do
+    encoded =
+      metadata
+      |> Enum.map(fn {k, v} -> "#{serialize(k)}=#{serialize(v)}" end)
+      |> Enum.join(",")
+    "HoundMetadata (#{encoded})"
+  end
+
+  @doc """
+  Extracts and parses the metadata contained in a user agent string.
+  If the user agent does not contain any metadata, an empty map is returned.
+  """
+  @spec parse(String.t) :: %{String.t => String.t}
+  def extract(str) do
+    ua_last_part = str |> String.split("/") |> List.last
+    case Regex.run(@extract_regexp, ua_last_part) do
+      [_, metadata] -> parse(metadata)
+      _             -> %{}
+    end
+  end
+
+  defp parse(metadata) do
+    metadata
+    |> String.split(",")
+    |> Enum.map(fn str ->
+      [k, v] = String.split(str, "=")
+      {deserialize(k), deserialize(v)}
+    end)
+    |> Enum.into(%{})
+  end
+
+  defp serialize(value) do
+    value
+    |> :erlang.term_to_binary
+    |> Base.encode64
+    |> String.rstrip(?=)
```

Comment:
Nevermind, I see why we are stripping. It should not be necessary to strip it if we encode the whole map.